[#1786] Trackevents no longer overlap but inhabit the track below instea...

dseif opened the pull request:
...d or create a new one

dseif committed:
[#1786] Trackevents no longer overlap but inhabit the track below instead or create a new one

[t1786] very simple ghosting working

[t1786] multi-level ghosting operational

[t1786] working without ghost tracks (for now)

[t1786] ghost track insertion

1. Added TrackEventDragManager to handle all of the mess associated with dragging and dropping with ghosts.
2. Added another trackevent on the middle track of the test template through saved-data.json.
3. Fixed dragndrop to be over-prepared for situations where draggables move in the dom tree immediately after events.
4. Swapped TrackEvent out of track view element during drag, and back in again on drop.

[t1786] Fixed up track handles and track ids

One bug in scrollbars remains!

[t1786] fixed scrollbars

[t1786] lint cleanup

ScottDowne commented:
/botio preview

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/693dbf31c620176/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/693dbf31c620176/output.txt

Total script time: 0.75 mins

Server is running on http://pr683.butyr.org/templates/test.html

ScottDowne commented:
/botio preview

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/4e1dc1bcefda0cf/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/4e1dc1bcefda0cf/output.txt

Total script time: 1.07 mins

Server is running on http://pr683.butyr.org/templates/test.html

secretrobotron committed:
[t1786] bug fix: trackevent reversion

secretrobotron committed:
[t1786] Removed bad try/catch statement

secretrobotron committed:
[t1786] Review fixes

1. Fixed tests by adding check for trackEventViews being on the DOM (0-width rects).
2. Squashed a couple of trackeventupdate leaving behind ghosts bugs.
3. Scrollbar fix.

secretrobotron committed:
[t1786] some fixes for dropping plugins and tes

secretrobotron committed:
[t1786] bug stomping

1. Added a stopPropagation function to events for stopping events from continuing down the listener chain.
2. trackEventUpdated in trackevent-drag-manager returns whether or not a new event was created.
3. track-container, which listens for trackeventupdated, calls stopPropagation if a new event was created (ghost).

dseif committed:
Merge with master

dseif committed:
[#1786] Trackhandles now reorder properly

dseif committed:
Merge branch 'master' into t1786

ScottDowne commented:
/botio preview

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/5dec80232031fe5/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/5dec80232031fe5/output.txt

Total script time: 1.29 mins

Server is running on http://pr683.butyr.org/templates/basic/
Run the Test Suite: http://pr683.butyr.org/test/

ScottDowne commented on src/core/track.js:
```diff
@@ -50,7 +48,9 @@ define( [
     this.updateTrackEvents = function() {
       for ( var i = 0, l = _trackEvents.length; i < l; i++ ) {
-        _trackEvents[ i ].update();
+        if ( !_trackEvents[ i ].view.ghost ) {
```
I wonder if it makes more sense to put this if inside the trackEvent.update. And if ghost, return early.

ScottDowne replied:
Yeah, inside trackEvent.update, check _isGhost.

ScottDowne commented on src/core/views/track-view.js:
```diff
@@ -24,43 +24,58 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
-
-          var start,
+          var draggableType = droppedElement.getAttribute( "data-butter-draggable-type" ),
+              start,
```
Why is this in two seperate var blocks?

ScottDowne commented on src/core/views/track-view.js:
```diff
@@ -127,21 +147,83 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
+        }
+      }
+      return null;
+    };
```
These two functions, findOverlappingTrackEvent and findOverlappingTrackEventFromRect are far too similar.

Either remove findOverlappingTrackEvent outright. Or, update findOverlappingTrackEvent to simply call findOverlappingTrackEventFromRect, and pass in the rect.

I already see a bug that was fixed in findOverlappingTrackEvent that was not fixed in findOverlappingTrackEventFromRect.

in findOverlappingTrackEvent
if ( trackEventView !== otherTrackEventView && !otherTrackEventView.dragging ) {

and

findOverlappingTrackEvent
if ( !otherTrackEventView.dragging ) {

Should probably want to pull the findOverlappingTrackEvent version in to findOverlappingTrackEvent. Seems more complete.

ScottDowne commented:
/botio preview

dseif committed:
[#1786] Fixed review comments

humphd commented:
/botio check

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @humphd received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/1943d697ca7614c/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/1943d697ca7614c/output.txt

Total script time: 0.93 mins

humphd commented on src/core/eventmanager.js:
```diff
@@ -132,7 +141,7 @@ define( [], function(){
         type: eventName.type,
         target: eventName.target,
-        data: eventName.data
+        data: eventName.data,
```
This comma is wrong.

humphd commented on src/core/track.js:
```diff
@@ -21,16 +19,16 @@ define( [
 
     _this._media = null;
     _this.order = 0;
+    _this.ghost = null;
```
Please add a comment to explain what this variable is for, since the name doesn't really indicate much.

humphd commented on src/core/trackevent.js:
```diff
@@ -41,7 +41,7 @@ define( [
    *
    * @param {Object} options: Options for initialization. Can contain the properties type, name, and popcornOptions. If the popcornOptions property is specified, its contents will be used to initialize the plugin instance associated with this TrackEvent.
```
Please update this comment to reflect the fact that the options object is not the only thing passed in, and what isGhost means.

humphd commented on src/core/trackevent.js:
```diff
@@ -262,6 +270,21 @@ define( [
+      /*
+       * Property: isGhost
+       *
+       * Specifies whether this trackEvent is a ghost or not
```
And being a ghost means what?  Needs more info somewhere on the whole "ghost" concept.

humphd commented on src/core/views/track-view.js:
```diff
@@ -24,43 +24,57 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
+            trackEventView = dropped.data;
+            track = trackEventView.trackEvent.track;
+
+            //Only rearrange trackEvent if it was moved *onto* this track
```
Space after //

humphd commented on src/core/views/track-view.js:
```diff
@@ -127,21 +146,66 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
+    };
 
-  }; //TrackView
+    this.addTrackEventGhost = function( ghost ) {
```
Again, add some kind of overall discussion comment that explains what a ghost is.

humphd commented on src/core/views/trackevent-view.js:
```diff
@@ -56,6 +58,48 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop",
+     * when the trackevent is dropped
+     */
+    this.createGhost = function( track ) {
+      if ( _ghost ) {
```
Why throw?  Why not just return _ghost?

humphd commented on src/core/views/trackevent-view.js:
```diff
@@ -56,6 +58,48 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop",
+    /*
+     * Member: cleanupGhost
+     *
+     * Removes this trackEvents ghost and makes sure isGhost is set to false
```
trackEvent's (apostrophe)

humphd commented on src/timeline/ghost-track.js:
```diff
@@ -0,0 +1,79 @@
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [], function() {
```
Needs explanatory comment block at the front to indicate what this is all about.

humphd commented on src/timeline/trackevent-drag-manager.js:
```diff
@@ -0,0 +1,187 @@
+    this.trackEventDropped = function( trackEvent, newTrack, startTime ) {
+      var newTrackEvent,
+          oldTrack = trackEvent.track,
+          corn = trackEvent.popcornOptions,
```
change this to `popcornOptions` vs. corn.

humphd commented on src/util/dragndrop.js:
```diff
@@ -39,7 +39,7 @@ define([], function(){
   }
 
-  function onDragged( e ){
+  var onDragged = function( e ){
```
Why this change?  Functions are better for memory.

humphd commented:
/botio preview

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @humphd received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/6514bd2ce8152c0/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/6514bd2ce8152c0/output.txt

Total script time: 2.00 mins

humphd commented:
/botio preview

ghost commented:
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @humphd received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/8a39fbd4de3b827/output.txt

ghost commented:
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/8a39fbd4de3b827/output.txt

Total script time: 1.05 mins

Server is running on http://pr683.butyr.org/templates/basic/
Run the Test Suite: http://pr683.butyr.org/test/

dseif committed:
[#1786] Review fixes

jbuck commented:
Staged on master: https://github.com/mozilla/butter/commit/987ff9b32e7b597acd671792d4641881791827b3
